bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LastAddSyncedManager.java

```diff
@@ -0,0 +1,76 @@
+/*
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.client;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Handles LastAddSynced on the client side.
+ */
+class LastAddSyncedManager {
+
+    private final Map<Integer, Long> lastAddSyncedMap = new HashMap<>();
```

Comment:
I think using `Map<Integer, Long>` here is not efficient. 

- it has a lot of allocations, boxing and unboxing.
- it can be simplified with `long[]`
- most of the ensembles are pretty small (e.g. 3 or 5)

Reply:
Will do in next patch